sycl/doc/design/DeviceIf.md

```diff
@@ -338,4 +338,1011 @@ in that compilation phase.
+SPIR-V to SPIR-V translator that does the code specialization.  The output of
+this translator can then be sent to the backend UMD even if it does not support
+SPV\_INTEL\_spec\_conditional.
+
+
+## Alternate design for AOT targets
+
+This is an alternate design that we can consider for targets that are AOT
+compiled.  This alternate design is orthogonal to the alternate design
+presented above for non-AOT targets, so we could implement either alternate
+design on its own or both together.
+
+In any AOT design, we need to resolve the _Conditional Actions_ at some point
+in the AOT compilation.  In the primary design, this is done in
+`sycl-post-link`.  This has both advantages and disadvantages.  The
+disadvantage is that we may lose some LLVM IR optimizations because the
+optimization passes (which run before `sycl-post-link`) operate on the code
+before the _Conditional Actions_ are resolved.  As an example, the optimizer
+won't be able to inline the body of these _Conditional Action_ functions at
+their call sites.  We think this will be mitigated for Intel GPU targets,
+though, because `ocloc` (which runs after `sycl-post-link`) can perform these
+optimizations.
+
+As an alternative design, we could instead resolve the _Conditional Actions_ in
+an LLVM IR pass that happens much earlier in the pipeline.  This would allow
+the LLVM optimization passes to better optimize the code because the
+_Conditional Action_ functions will either be direct calls or they will be
+entirely eliminated by the time the optimizer passes run.  If we do this, it
+makes sense to resolve the _Conditional Actions_ sometime before the
+[SYCLPropagateAspectsUsagePass][3] runs.  The algorithm would be similar to
+what `sycl-post-link` does in the primary design:
+
+* Iterate over each call to a _Conditional Caller_ function.  These calls have
+  the following form:
+
+  ```
+  call void @call_if_on_device_conditionally_GUID_N(
+              @CallableXXX, %callablethis, N1, N2, ...)
+  ```
+
+* Using the device configuration file, determine if the _Condition
+  Expression_ is true for this target.
+
+* If the condition is true, change the call site into a direct call to the
+  _Conditional Action_ function like so:
+
+  ```
+  call void @CallableXXX(%callablethis)
+  ```
+
+* If the condition is false, remove the call site entirely.
+
+The SYCLPropagateAspectsUsagePass will then propagate aspects used by the
+remaining _Conditional Actions_ up to the kernel functions as
+`!sycl_used_aspects` metadata (i.e. unconditionally used aspects).
+
+The disadvantage to this design is that it increases compilation time when
+there are multiple targets.  Once the _Conditional Actions_ are resolved,
+the LLVM IR is now specialized for one particular AOT target.  If the user has
+asked to compile for multiple targets, we need to split the IR prior to
+resolving the _Conditional Actions_ and run the LLVM IR pipeline separately for
+each copy of the IR.
```

Comment:
I think that this is what we currently do anyway: each AOT target (like `spir64_gen` or `spir64_x86_64`) causes a separate invocation of FE and thus separate optimization pipeline invocation. so, this won't be worse than what we have today.

But at the same time, we currently have only a few AOT targets because they are generic and do not represent a particular device. Once we switch to specialized targets like `spir64_intel_skl` running FE separately for each of them would be a disaster and we better not lock-in ourselves into such situation.

What we could do is a single FE invocation and then several `opt` invocations for each target to at least somehow mitigate this.

Alternatively, we could just leave a single IR for all targets, but be more tricky about the IR we emit: instead of dropping bodies of `call_if_on_device_conditionallyXXX` we could replace their bodies to get the following:
```
define void @call_if_on_device_conditionallyXXX(%callablethis, %n1, %n2, ...) {
  %bool = __sycl_builtin_if_on_device.mangling(%n1, %n2, ...)
  br %bool, label %do_call, label %exit
do_call:
  call void @CallableXXX(%callablethis)
exit:
  ret void
}
```

This would be an alternative transformation which the new pass does for AOT targets. Yes, it won't allow to full optimizations and may still limit some inlining, but it is better than nothing and allows us to have unified IR for all AOT targets in the optimization pipeline. Later, in `sycl-post-link` we would resolve that `__sycl_builtin_if_on_device.*` into concrete boolean value based on the target device and run some DCE to cleanup dead branches.

Reply:
> Once we switch to specialized targets like `spir64_intel_skl` running FE separately for each of them would be a disaster and we better not lock-in ourselves into such situation.

Yes, this is what I was worried about when describing the disadvantages of this design.

> What we could do is a single FE invocation and then several `opt` invocations for each target to at least somehow mitigate this.

Yes, this is what I meant by "split the IR prior to resolving the _Conditional Actions_ and run the LLVM IR pipeline separately for each copy of the IR".

> Alternatively, we could just leave a single IR for all targets, but be more tricky about the IR we emit: instead of dropping bodies of `call_if_on_device_conditionallyXXX` we could replace their bodies to get the following

I'm not sure this is a good idea.  I'm worried that the body of `@CallableXXX` will get inlined and then some instructions from its body will be lifted above the `br %bool`.  I saw some cases like this when experimenting with a solution like this.  The nice thing about the primary design is that there is no call to `@CallableXXX` in the IR module, so it cannot possibly be inlined, thus avoiding the possibility of "leaking" any of its instructions above the condition check.

My current thinking is that:

* We should use this alternate design for Nvidia targets because these targets run the FE separately for each target anyway.
* We should use the primary design for Intel GPU targets and see if `ocloc` does a good job of inlining the calls to `@CallableXXX` and optimizing the result.

If `ocloc` optimizations are not good, we could try the alternate design, where we split the IR and run optimization passes separately for each Intel GPU target.